### PR TITLE
Add ncclCustomCollective to interpreter SCCL XML

### PR DESCRIPTION
**Pull request by abhijangda**

### CODE REVIEW

File: src/collectives/device/custom_collective.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ ************************************************************************/
+
+ #include "sccl_interpreter.h"
+ 
+template<class FUNC, typename T, int UNROLL>
+class ncclFunction<ncclFuncCustomCollective, NCCL_ALGO_SCCL, NCCL_PROTO_SIMPLE, FUNC, T, UNROLL> {
+  public:
+    __device__ void run(struct ncclWorkElem* args) {
+      scclFunctionSimple<FUNC, T, UNROLL> scclfunc;
+      scclfunc.run(args, args->comm->nRanks);
```
Comment (saeedmaleki) [inline src/collectives/device/custom_collective.h]: why args->comm->nRanks?

File: src/graph/topo.cc [hunk 2]
```diff
@@ -642,8 +642,7 @@ ncclResult_t scclCheckBufferBounds(int bufferType, int offset, int nInputChunks,
   return ncclSuccess;
 }
 
-ncclResult_t scclGetAlgoFromXMLAndSetComm(struct ncclComm* comm) {
-  char* str = getenv("SCCL_XML_FILE");
+ncclResult_t scclGetAlgoFromXMLAndSetComm(struct ncclComm* comm, const char* str) {
```
Comment (saeedmaleki) [inline src/graph/topo.cc:645]: So does this mean we can take multiple XML files?

File: src/graph/topo.cc [hunk 3]
```diff
@@ -667,6 +666,25 @@ ncclResult_t scclGetAlgoFromXMLAndSetComm(struct ncclComm* comm) {
   NCCLCHECK(xmlGetAttrInt(topNode, "nchunksperloop", &nchunksPerLoop));
   int globalNChannels;
   NCCLCHECK(xmlGetAttrInt(topNode, "nchannels", &globalNChannels));
+
+  const char* redop;
+  NCCLCHECK(xmlGetAttrStr(topNode, "redop", &redop));
+  if (strcmp(redop, "sum") == 0){
```
Comment (saeedmaleki) [inline src/graph/topo.cc:672]: why is the FUNC not enough? in what kinda of scenario we might need multiple?

File: src/graph/tuning.cc [hunk 4]
```diff
@@ -117,24 +117,26 @@ ncclResult_t ncclTopoTuneModel(struct ncclComm* comm, int minCompCap, int maxCom
     // Skipping on SCCL algorithms here since SCCL tune the algorithm in the synthesizer.
 
     for (int a=0; a<NCCL_NUM_ALGORITHMS; a++) {
-      if (coll == ncclFuncAllToAll || a == NCCL_ALGO_SCCL) {
-        // SCCL algorithm has hooks for AllToAll, AllGather, AllReduce and ReduceScatter
```
Comment (saeedmaleki) [inline src/graph/tuning.cc:121]: why are you taking out allreduce?

File: src/graph/tuning.cc [hunk 5]
```diff
@@ -117,24 +117,26 @@ ncclResult_t ncclTopoTuneModel(struct ncclComm* comm, int minCompCap, int maxCom
...
+      if (coll == ncclFuncAllToAll || coll == ncclFuncCustomCollective || a == NCCL_ALGO_SCCL) {
+        // SCCL algorithm has hooks for AllToAll, AllGather and ReduceScatter
         // SCCL algorithm is dynamic and busBw/latency can only be determined by the input XML algorithm. An analysis will be added later.
         for (int p=0; p<NCCL_NUM_PROTOCOLS; p++) {
-          if ((coll == ncclFuncAllToAll || coll == ncclFuncAllGather || coll == ncclFuncReduceScatter || coll == ncclFuncAllReduce) && a == NCCL_ALGO_SCCL && p == comm->scclAlgo.protocol){
```
Comment (saeedmaleki) [inline src/graph/tuning.cc]: same question

File: src/graph/tuning.cc [hunk 6]
```diff
@@ -117,24 +117,26 @@ ncclResult_t ncclTopoTuneModel(struct ncclComm* comm, int minCompCap, int maxCom
...
             comm->latencies[coll][a][p] = 0.0;
           }
         }
         continue;
       }
 
+      
       if (coll != ncclFuncAllReduce && a != NCCL_ALGO_RING) continue;
-
+      if (coll == ncclFuncCustomCollective) continue;
```
Comment (saeedmaleki) [inline src/graph/tuning.cc]: this should not be necessary based on tuning parameters set earlier, right? There is only one customCollective algorithm and that only one should have a > 0  bw

File: src/group.cc [hunk 7]
```diff
@@ -69,11 +70,11 @@ thread_local struct ncclAsyncArgs ncclGroupArgs[MAX_ASYNC_OPS];
 
 void* ncclAsyncThreadMain(void* args_) {
   struct ncclAsyncArgs* args = (struct ncclAsyncArgs*)args_;
-  NCCLCHECKTHREAD(args->init.func(args->init.newcomm, args->init.ndev, args->init.commId, args->init.myrank, args->init.cudaDev));
+  NCCLCHECKTHREAD(args->init.func(args->init.newcomm, args->init.ndev, args->init.commId, args->init.myrank, args->init.cudaDev, args->init.xmlFile));
   return args;
 }
 
-ncclResult_t ncclAsyncInit(ncclInitFunc_t func, ncclComm_t* newcomm, int ndev, ncclUniqueId commId, int myrank, int cudaDev) {
+ncclResult_t ncclAsyncInit(ncclInitFunc_t func, ncclComm_t* newcomm, int ndev, ncclUniqueId commId, int myrank, int cudaDev, const char* xmlFile) {
```
Comment (saeedmaleki) [inline src/group.cc]: some comments in here would be useful.

File: src/group.cc [hunk 8]
```diff
@@ -69,11 +70,11 @@ thread_local struct ncclAsyncArgs ncclGroupArgs[MAX_ASYNC_OPS];
 
 void* ncclAsyncThreadMain(void* args_) {
   struct ncclAsyncArgs* args = (struct ncclAsyncArgs*)args_;
-  NCCLCHECKTHREAD(args->init.func(args->init.newcomm, args->init.ndev, args->init.commId, args->init.myrank, args->init.cudaDev));
+  NCCLCHECKTHREAD(args->init.func(args->init.newcomm, args->init.ndev, args->init.commId, args->init.myrank, args->init.cudaDev, args->init.xmlFile));
```
Comment (saeedmaleki) [inline src/group.cc]: some comments please.

File: src/group.cc [hunk 9]
```diff
@@ -31,6 +31,7 @@ struct ncclInitArgs {
   int ndev;
   ncclUniqueId commId;
   int myrank;
+  char xmlFile[1024];
```
Comment (saeedmaleki) [inline src/group.cc]: Why is there a xmlfile in group.cc? What is the one in comm.h then?

File: src/include/comm.h [hunk 10]
```diff
@@ -152,6 +152,9 @@ struct ncclComm {
   struct ncclP2Plist* p2pRecvs;
   int p2pSendCount;
   int p2pRecvCount;
+
+  //Custom Collective information
+  char xmlFile[1024];
```
Comment (saeedmaleki) [inline src/include/comm.h]: Shouldn't this be remove then now that you have one in group.cc?

File: src/include/devcomm.h [hunk 11]
```diff
@@ -195,6 +196,9 @@ struct scclAlgorithm {
   // declaration for scratchBuffer. This is only to be accessed by the host
   size_t scratchBufferSize;
   void* scratchBuffer;
+  //Reduction Operator. If the algorithm performs reduction it will specify the reduction operator.
+  //If the algorithm do not perform reduction, its reduction operator is considered as ncclSum.
+  ncclRedOp_t redOp;
```
Comment (saeedmaleki) [inline src/include/devcomm.h:201]: yeah, still not sure why we need this. FUNC will apply the same reduction operation all across, right?

File: src/init.cc [hunk 12]
```diff
@@ -210,7 +210,7 @@ static ncclResult_t commFree(ncclComm_t comm) {
   return ncclSuccess;
 }
 
-static ncclResult_t commAlloc(ncclComm_t* comret, int ndev, int rank) {
+static ncclResult_t commAlloc(ncclComm_t* comret, int ndev, int rank, const char* xmlFile) {
```
Comment (saeedmaleki) [inline src/init.cc]: nice, does this mean we can call commAlloc multiple times?

File: src/init.cc [hunk 13]
```diff
@@ -864,10 +867,15 @@ static ncclResult_t initTransportsRank(struct ncclComm* comm, ncclUniqueId* comm
 
   // Compute nChannels per peer for p2p
   NCCLCHECK(ncclTopoComputeP2pChannels(comm));
-
+
```
Comment (saeedmaleki) [inline src/init.cc]: remove.

File: src/init.cc [hunk 14]
```diff
@@ -976,7 +984,16 @@ ncclResult_t ncclCommInitRank(ncclComm_t* newcomm, int nranks, ncclUniqueId comm
   NVTX3_FUNC_RANGE_IN(nccl_domain);
   int cudaDev;
   CUDACHECK(cudaGetDevice(&cudaDev));
-  NCCLCHECK(ncclCommInitRankDev(newcomm, nranks, commId, myrank, cudaDev));
+  NCCLCHECK(ncclCommInitRankDev(newcomm, nranks, commId, myrank, cudaDev, nullptr));
+  return ncclSuccess;
+}
+
+NCCL_API(ncclResult_t, ncclCommInitRankWithScclXML, ncclComm_t* newcomm, int nranks, ncclUniqueId commId, int myrank, const char* xmlFile);
+ncclResult_t ncclCommInitRankWithScclXML(ncclComm_t* newcomm, int nranks, ncclUniqueId commId, int myrank, const char* xmlFile) {
```
Comment (saeedmaleki) [inline src/init.cc]: Please add a comment for each newly introduced function

File: src/misc/argcheck.cc [hunk 15]
```diff
@@ -45,12 +45,12 @@ ncclResult_t ArgsCheck(struct ncclInfo* info) {
   }
   // Type is OK, compute nbytes. Convert Allgather/Broadcast/P2P calls to chars.
   info->nBytes = info->count * ncclTypeSize(info->datatype);
-  if (info->coll == ncclFuncAllGather || info->coll == ncclFuncBroadcast || info->coll == ncclFuncAllToAll) {
+  if (info->coll == ncclFuncAllGather || info->coll == ncclFuncBroadcast || info->coll == ncclFuncAllToAll) { //|| info->coll == ncclFuncCustomCollective
```
Comment (saeedmaleki) [inline src/misc/argcheck.cc]: what's going on here?

File: src/misc/argcheck.cc [hunk 16]
```diff
@@ -45,12 +45,12 @@ ncclResult_t ArgsCheck(struct ncclInfo* info) {
...
     info->count = info->nBytes;
     info->datatype = ncclInt8;
   }
   
-  if (info->coll == ncclFuncAllToAll || info->coll == ncclFuncAllGather || info->coll == ncclFuncReduceScatter) info->nBytes *= info->comm->nRanks; // count is per rank
+  if (info->coll == ncclFuncAllToAll || info->coll == ncclFuncAllGather || info->coll == ncclFuncReduceScatter) info->nBytes *= info->comm->nRanks; // count is per rank || info->coll == ncclFuncCustomCollective
```
Comment (saeedmaleki) [inline src/misc/argcheck.cc]: that is a good question, shouldn't we make it *=1? I may have a collective that is not divisible by nranks, right?

**Review by saeedmaleki — COMMENTED** [review]
Nice job! Documentation is the most important part for a PR. I really want to understand how you handle dynamic XML files.

File: src/graph/topo.cc (hunk 3 above)
Comment (abhijangda) [reply src/graph/topo.cc:672]: All ncclKernels and ncclFuncs take a FUNC argument, even those that do not use FUNC like AllGather.
By default in NCCL these kernels take Sum as the FUNC argument value. 
Similarly, CustomCollective kernels have SUM, Prod, and other as reduction operators.
This operator can be passed through ncclCustomCollective or through XML. I think passing it through XML is cleaner.

File: src/graph/tuning.cc (hunk 4 above)
Comment (abhijangda) [reply src/graph/tuning.cc:121]: Weird. I think git did not give me a merge conflict for these.

**Comment by abhijangda** [comment]
I fixed all of @saeedmaleki 's comments

File: src/init.cc (hunk 14 above)
Comment (abhijangda) [reply src/init.cc]: They are added in nccl.h.in for the API functions.

File: src/graph/topo.cc (hunk 2 above)
Comment (abhijangda) [reply src/graph/topo.cc:645]: Not yet.

**Pull request merged**